crates/biome_js_analyze/src/lint/nursery/no_nested_component_definitions.rs

```diff
@@ -0,0 +1,279 @@
+    /// function ParentComponent() {
+    ///   const MemoizedChild = memo(() => {
+    ///     return <div>Hello</div>;
+    ///   });
+    ///
+    ///   return <MemoizedChild />;
+    /// }
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// Component is defined outside other components:
+    /// ```jsx
+    /// function ChildComponent() {
+    ///   return <div>Hello</div>;
+    /// }
+    ///
+    /// function ParentComponent() {
+    ///   return <ChildComponent />;
+    /// }
+    /// ```
+    ///
+    /// ## Correct approaches
+    ///
+    /// 1. Move the component definition outside:
+    ///    ```jsx
+    ///    function ChildComponent() {
+    ///      return <div>Hello</div>;
+    ///    }
+    ///
+    ///    function ParentComponent() {
+    ///      return <ChildComponent />;
+    ///    }
+    ///    ```
+    ///
+    /// 2. Pass components as props:
+    ///    ```jsx
+    ///    function ParentComponent({ CustomComponent }) {
+    ///      return <CustomComponent />;
+    ///    }
+    ///    ```
+    ///
+    /// 3. Use React's Children API:
+    ///    ```jsx
+    ///    function ParentComponent({ children }) {
+    ///      return <div>{children}</div>;
+    ///    }
+    ///    ```
+    pub NoNestedComponentDefinitions {
+        version: "next",
+        name: "noNestedComponentDefinitions",
+        language: "js",
+        sources: &[RuleSource::EslintReact("no-nested-components")],
+        recommended: false,
+    }
+}
+
+impl Rule for NoNestedComponentDefinitions {
+    type Query = Ast<AnyJsFunction>;
+    type State = RuleState;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        if let Some(component_name) = get_function_component_info(node) {
+            if let Some(parent_component_name) = node
+                .syntax()
+                .ancestors()
+                .skip_while(|ancestor| ancestor.eq(node.syntax()))
+                .find_map(|ancestor| {
+                    AnyJsFunction::cast(ancestor).and_then(|n| get_function_component_info(&n))
+                })
+            {
+                return Some(RuleState {
+                    component_name,
+                    parent_component_name,
+                });
+            }
+        };
+        None
+    }
+
+    fn diagnostic(_ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                state.component_name.text_range(),
+                markup! {
+                    "Components should not be defined inside other components."
+                },
+            )
+            .detail(
+                state.parent_component_name.text_range(),
+                markup! {
+                    "Move it outside of the parent component or pass it as a prop."
+                },
+            ),
```

Comment:
We can add why the rule occurs here, https://biomejs.dev/linter/#rule-pillars
The explanation maybe useful here.

> Component definitions inside other components cause them to be recreated on every render, which can lead to performance issues and unexpected behavior.

Reply:
Thanks, done!